Type safety for the VisualStudio Backend(s)

dcbaker opened the pull request:
~~This gets most of the VS backend type safe, with the exception of some of the object extraction stuff where there seems to be a real potential for things to explode.~~

This gets all of the VS Backends type safe. Thanks @bonzini for for help with the object stuff.

This turned up a lot of issues in the build module and in the base `backend` module.

It should be possible to merge pieces of this series incrementally.

It should be noted that this is untested on the vs2010 backend. I expect to find issues at runtime.

dcbaker committed:
options: Add a TypeAlias for option values

This gives us a simpler way to annotate things returning an option
value, and gives us an easy single place to change that will affect
everywhere if the option value types are changed.

dcbaker committed:
build: fix typing of `Target.get_option`

Which can return any of `ElementaryOptionValues`, but is currently typed
as if it only returns `str | int | bool`.

dcbaker committed:
build: fix typing of get_target_dir_relative_to

Which should be `Target | CustomTargetIndex`, not just `Target`

dcbaker committed:
build: Add a property for depend_files to CustomTargetIndex

So that it can meet it's interface requirements.

dcbaker committed:
backend/backends: fix get_target_depend_files typing

Which may be passed a BuildTarget, CustomTarget, CustomTargetIndex, or
GeneratedList

This requires using `get_subdir()` because `CustomTargetIndex` doesn't
have a `subdir` attribute.

dcbaker committed:
backend/vs2010: Add the majority of the missing type annotations

This is mostly easy, simple stuff to get out of the way before tackling
the harder issues

dcbaker committed:
backend/backends: Fix annotations of relpath

Which can actually be a `PathLike[str] | str`, not just `str`

dcbaker committed:
backend/backends: Remove uses of OrderedDict that can just be Dict

dcbaker committed:
backend/backends: fix typing of as_meson_exe_cmdline and get_executable_serialisation

Which should take `Iterable[BuildTarget | CustomTarget | CustomTargetIndex]`, not
just `List[BuildTarget]`.

dcbaker committed:
backend/vs2010: fix typing of pch code

This is a little more invasive since we're currently using lists where
we should be using tuples. I've added a TypeAlias to make this code
clearer and cleaner.

dcbaker committed:
interpreter: catch '@DEPFILE@' and raise an exception

If this gets to the backend it will cause an uncaught exception, so
let's catch it here and just give a real error.

dcbaker committed:
backend/backends: Fix eval_custom_target with RunTarget

At least in the vs backend this will get called, and it will fail. So
not only do we need to update the type annotations, but we also need to
fix the implementaiton.

dcbaker committed:
backend/backends: tighten types in get_target_filename_for_linking

It's really only `BuildTarget | CustomTarget | CustomTargetIndex` that
can be used here, so let's do that correctly.

Also, use only `target.import_filename` since mypy doesn't know about
the getter function.

dcbaker committed:
build: add get_option to `CustomTargetIndex`

Which it needs in the vs2010 backend

dcbaker committed:
build: Add get_basename to CustomTargetIndex

Guess what the vs2010 backend can hit?

dcbaker committed:
backend/backends: fix build_target_link_arguments typing

This is actually an Iterable, since we pass both lists and OrderedSets
in, but also it's not `Target`, it's `BuildTargetTypes`

dcbaker committed:
backend/backends: fix typing with generate_unity_files

which needs to take `Iterable[File | str]` not `str`

dcbaker committed:
backend/vs2010: fix a lot of `str`, `File`, and `File | str` issues

This gets the vs2010 backend completely mypy compliant

dcbaker committed:
backend/vs2017: fully typesafe

dcbaker committed:
backend/vs2019: fully type safe

dcbaker committed:
backend/vs2022: fully type safe

dcbaker committed:
backend/vs2010: stop using dict.keys()

It's almost never what you want

dcbaker committed:
backend/vs2010: stop using `len` to test for container emptiness

dcbaker committed:
build: type BuildTarget.link_depends

dcbaker committed:
utils/universal: make classify Unity Sources generic

This means it can be either `str`, `File`, or `File | str`, which gives
us the value of knowing that if the input is T, the output is
Dict[Compiler, List[T]]`

dcbaker committed:
backend/vs2010: use itertools.chain instead of adding lists

The latter of which creates a new list with references to all of the
objects in the inputs. In our case that's wasted as we just iterate
anyway.

bonzini commented:
This seems to do what's missing for VS2010 extracted objects

```
diff --git a/mesonbuild/backend/vs2010backend.py b/mesonbuild/backend/vs2010backend.py
index 3c85d7983..cf4107677 100644
--- a/mesonbuild/backend/vs2010backend.py
+++ b/mesonbuild/backend/vs2010backend.py
@@ -32,6 +32,7 @@ if T.TYPE_CHECKING:
     from ..arglist import CompilerArgs
     from ..compilers import Compiler
     from ..interpreter import Interpreter
+    from ..mesonlib import FileOrString
 
     Project: TypeAlias = T.Tuple[str, PurePath, str, MachineChoice]
     PchSources: TypeAlias = T.Tuple[str, T.Optional[str], str, T.Optional[str]]
@@ -1514,6 +1515,8 @@ class Vs2010Backend(backends.Backend):
                     # Expand our object lists manually if we are on pre-Visual Studio 2015 Update 2
                     l = t.extract_all_objects(False)
 
+                    src: FileOrString
+
                     # Unfortunately, we can't use self.object_filename_from_source()
                     for gen in l.genlist:
                         for src in gen.get_outputs():
@@ -1529,7 +1532,8 @@ class Vs2010Backend(backends.Backend):
                             rel_obj = self.object_filename_from_source(t, compiler, src, target_private_dir)
                             extra_link_args.append(rel_obj)
 
-                    extra_link_args.extend(self.flatten_object_list(t)[0])
+                    if isinstance(t, build.BuildTarget):
+                        extra_link_args.extend(self.flatten_object_list(t)[0])
                 else:
                     # /WHOLEARCHIVE:foo must go into AdditionalOptions
                     extra_link_args += compiler.get_link_whole_for([linkname])
@@ -1824,7 +1828,6 @@ class Vs2010Backend(backends.Backend):
 
         additional_objects: T.List[str] = []
         for o in self.flatten_object_list(target, proj_to_build_root)[0]:
-            assert isinstance(o, str)
             additional_objects.append(o)
         for o in custom_objs:
             additional_objects.append(o)
diff --git a/mesonbuild/build.py b/mesonbuild/build.py
index 388f4877b..d4580e57f 100644
--- a/mesonbuild/build.py
+++ b/mesonbuild/build.py
@@ -2616,6 +2616,9 @@ class CustomTargetBase:
 
     rust_crate_type = ''
 
+    def extract_all_objects(self, recursive: bool = False) -> ExtractedObjects:
+        return ExtractedObjects(self, [], self.get_outputs(), [], recursive)
+
     def get_dependencies_recurse(self, result: OrderedSet[BuildTargetTypes], include_internals: bool = True) -> None:
         pass
 
@@ -2810,9 +2813,6 @@ class CustomTarget(Target, CustomTargetBase, CommandBase):
             return False
         return CustomTargetIndex(self, self.outputs[0]).is_internal()
 
-    def extract_all_objects(self) -> T.List[T.Union[str, 'ExtractedObjects']]:
-        return self.get_outputs()
-
     def type_suffix(self):
         return "@cus"
 
@@ -3084,9 +3084,6 @@ class CustomTargetIndex(CustomTargetBase, HoldableObject):
         suf = os.path.splitext(self.output)[-1]
         return suf in {'.a', '.lib'} and not self.should_install()
 
-    def extract_all_objects(self) -> T.List[T.Union[str, 'ExtractedObjects']]:
-        return self.target.extract_all_objects()
-
     def get_custom_install_dir(self) -> T.List[T.Union[str, Literal[False]]]:
         return self.target.get_custom_install_dir()
 
```

I think the build.py parts can be placed in front of the whole series, whereas the vs2010 parts need to be split and squashed.

There is coverage in `test cases/common/208 link custom`, which passes on Linux.

dcbaker commented:
@bonzini: the extract_all_objects methods get used all over the place, in the Interpreter, the Ninja Backend, and within build itself. I don't think we can drop those.

bonzini commented:
They aren't dropped, just unified to CustomTargetBase. The CustomTargetIndex case was wrong in that it returned all objects instead of just the required index; and they had the wrong type because they could return `str`.

dcbaker committed:
build: unify and correct CustomTarget derived extract_all_objects

There are a couple of issues with the current implementation:
1. The API doesn't match the BuildTarget on
2. The CustomTargetIndex version returns all of the CustomTarget's
   objects, not just the one for that index

To fix that the extract_all_objects method is moved to the
CustomTargetBase class, and extended to implement the same API as the
BuildTarget implementaiton.

This also finds a bunch of previously incorrectly typed (and thus
unhandled) cases in the Backend base class around ExtractedObjects from
CustomTargets

Co-Authored-by: Paolo Bonzini <pbonzini@redhat.com>

dcbaker commented:
Ah, I see what you did now. Okay, yeah, plus a little work in the ExtractedObjects class and we can make that work.

bruchar1 commented on mesonbuild/build.py:
```diff
@@ -3090,6 +3090,9 @@ def extract_all_objects(self) -> T.List[T.Union[str, 'ExtractedObjects']]:
     def get_custom_install_dir(self) -> T.List[T.Union[str, Literal[False]]]:
         return self.target.get_custom_install_dir()
 
+    def get_option(self, key: 'OptionKey') -> T.Union[str, int, bool, T.List[str]]:
```
Why is OptionKey quoted here ?

dcbaker replied:
Copy and past. Fixed in the latest version

bruchar1 commented on mesonbuild/backend/vs2010backend.py:
```diff
@@ -356,6 +356,7 @@ def get_target_deps(self, t: T.Dict[T.Any, T.Union[build.Target, build.CustomTar
                     if isinstance(ldep, build.CustomTargetIndex):
                         all_deps[ldep.get_id()] = ldep.target
                     elif isinstance(ldep, File):
+                        # XXX: this this true? What if the file `is_built`?
```
Is there an action to take here?

dcbaker replied:
I mean, probably. But I don't have a windows machine handy to find out what happens if a generated file is used for a link_depends and fix it.

bruchar1 replied:
Ok. I will try to look at this (maybe not in a near future, but someday...)

dcbaker replied:
I did reduce it from an `XXX comment to just a comment. along with fixing the `this this` :)

bruchar1 reviewed (commented):
I don't see any fancy stuff after the commit "[Fix a lot of low hanging type issues]". Is the description of that commit still relevant?
